### PR TITLE
Expose presence_penalty, frequency_penalty, and per-penalty context_size on the server API

### LINKED ISSUE #1021: GLM-OCR: /chat/completions produces repetition loops on prose; repetition_penalty does not prevent them
## Summary

When OCR'ing a scanned page of flowing English prose through the mlx-vlm server (`/chat/completions`) with `mlx-community/GLM-OCR-bf16`, the model locks into a repetition loop after the first ~500 correct characters and re-emits the same short paragraph dozens of times until `max_tokens` truncates. Setting `repetition_penalty` on the request does not prevent this.

The same source image processed by the same model weights via Ollama (GGUF quantisation, llama.cpp sampler) produces clean, coherent OCR with no loops — so the model itself and image understanding are fine; the issue is isolated to mlx-vlm's generation / sampling path.

## Environment

- mlx-vlm 0.4.4 (installed from `git+https://github.com/Blaizzy/mlx-vlm.git`)
- Python 3.12.13 (Homebrew, macOS arm64)
- macOS Darwin 24.6.0, Apple Silicon Mac Mini
- torch 2.11.0 + torchvision 0.26.0 installed (required for `GlmOcrProcessor` via closed issue #886)
- Model: `mlx-community/GLM-OCR-bf16` (also reproduces with `EZCon/GLM-OCR-mlx`)

## Reproduction

1. Start the server:
   ```
   python -m mlx_vlm.server --trust-remote-code --port 8080
   ```

2. Render a page of flowing English prose at 200 DPI to PNG (image-only scanned PDF, typical size ~300–500 KB). A public-domain Project Gutenberg scan reproduces the issue on any page with an all-caps section heading.

3. POST to `/chat/completions`:
   ```json
   {
     "model": "mlx-community/GLM-OCR-bf16",
     "messages": [{"role": "user", "content": [
       {"type": "image_url", "image_url": {"url": "data:image/png;base64,<...>"}},
       {"type": "text", "text": "OCR this image. Extract ALL text preserving the original formatting, paragraphs, tables, and formulas. Output only the extracted text."}
     ]}],
     "max_tokens": 4096,
     "temperature": 0,
     "repetition_penalty": 1.1
   }
   ```

4. Observe: the first ~500 characters of the response match the page content correctly. Then the model hits a short all-caps heading (e.g. a section header, ~3 words) and begins re-emitting the heading + the subsequent paragraph in a loop until `max_tokens`.

## Measured behaviour

Same 200-DPI rendering, single page of English prose, identical prompt:

| `repetition_penalty` | `temperature` | `top_p` | Output (chars) | Repeated-heading occurrences |
|---|---|---|---|---|
| *unset* | 0.0 | — | 33,688 | ~100 |
| 1.05 | 0.0 | — | 16,870 | 54 |
| 1.1 | 0.0 | — | 16,870 | 54 |
| 1.3 | 0.3 | — | 16,657 | 79 |
| 1.5 | 0.0 | — | 16,594 | 80 |
| 1.1 | 0.5 | 0.9 | 17,086 | 54 |

Ground truth (Ollama GGUF on same image, `glm-ocr:latest`, default sampler): **1,603 chars, heading appears once**. Complete correct transcription of the page.

Observations:
- `repetition_penalty` values from 1.05 to 1.5 do not break the loop on this content
- Increasing `repetition_penalty` from 1.1 → 1.5 at `temperature=0` actually *increased* repeat count (54 → 80)
- Temperature > 0 + `top_p` 0.9 did not help either
- The loop output consistently fills close to `max_tokens` (input ~5000 prompt tokens, output ~4096 generation tokens)

## Hypothesis

Either:
1. The `repetition_penalty` field on `/chat/completions` is accepted by the request model but not threaded into `stream_generate` / `generate_step` for VLM paths, or
2. It is applied but using a codebase convention (e.g. a context window length) that renders it ineffective for this model's tokenisation of repeated all-caps headings, or
3. GLM-OCR's chat template's pre-filled `<think></think>` interacts with sampling in a way that bypasses the penalty.

The counterfactual (Ollama/llama.cpp producing clean output from the same weights on the same image) strongly points at the mlx-vlm sampling path as the isolated cause.

## Related

- #886 (closed) — `GlmOcrProcessor` not loaded by AutoProcessor; same model, different failure mode. Resolved by installing torch+torchvision, but exposed this downstream sampler issue.
- #387 (closed, different model) — similar pattern of mlx-vlm generation diverging from llama.cpp.

## Why this matters

GLM-OCR via mlx-vlm is otherwise an excellent fit for Apple Silicon OCR pipelines (~2× throughput vs. Ollama on benchmarks). This sampler issue makes it unusable today for anyone processing long-form prose; structured content (receipts, forms, equations) appears to mask the problem because short repetitive headers are rarer there.

Happy to test patches against the same reproduction and report back. I'm also skimming the sampling code to see whether a small PR is feasible; will link if I find something.

Thanks for maintaining this project!

### PR DESCRIPTION
**Pull request by esaruoho**
Closes #1021.

## What

Surface `presence_penalty`, `frequency_penalty`, and per-penalty `*_context_size` fields on the `/chat/completions` and `/responses` endpoints, and forward them to `make_logits_processors`. `mlx_lm.sample_utils.make_logits_processors` already supports these — they were just never plumbed through.

No default behaviour changes; every new field defaults to `None`, so `make_logits_processors` falls back to its existing defaults if a caller omits them. The PR is purely additive surface area.

## Why

While debugging GLM-OCR-bf16 (#1021), I confirmed end-to-end with debug instrumentation that `repetition_penalty` *is* applied through the sampler — but the multiplicative penalty discounts a token *once per unique appearance in the recent window* (the Keskar 2019 formulation), not per occurrence. For VLMs whose logits are sharply peaked (OCR pages, classifier-style outputs), `repetition_penalty=1.1` (the typical text-LLM default) is below the threshold needed to move the argmax at `temperature=0`, and the model gets stuck looping a short heading + paragraph dozens of times until `max_tokens` cuts it off.

Empirical reproduction on a single page of English prose (image-only PDF, 200 DPI), `temperature=0`, `max_tokens=4096`:

| `repetition_penalty` | extra penalty | output (chars) | repeated heading occurrences |
|---|---|---|---|
| 1.1 | — | 16,870 | 54 |
| 1.5 | — | 16,594 | 80 |
| 2.0 | — | 1,188 | 2 ✅ |
| 1.1 | `frequency_penalty=0.5` | 4,444 | 14 |
| 1.1 | `frequency_penalty=1.0` | 4,164 | 7 |
| 1.1 | `presence_penalty=0.5` | 8,446 | 27 |

Ground-truth Ollama/llama.cpp on the same image: ~1,600 chars, 1 occurrence.

The point is not to argue for a particular default — it's that `frequency_penalty` and `presence_penalty` are the standard knobs for this failure mode, and they're already supported by the underlying sampler. Surfacing them lets callers tune for VLM/OCR workloads without resorting to extreme `repetition_penalty` values that suppress too aggressively.

## Changes

* **`mlx_vlm/server.py`** — extend `GenerationParams` with `repetition_context_size`, `presence_penalty`, `presence_context_size`, `frequency_penalty`, `frequency_context_size` (all `Optional[…] = None`), include them in `shared_generation_kwargs`. Updated the `repetition_penalty` description to flag the rp ≥ 2.0 caveat for VLMs. The new fields are inherited by both `ChatRequest` and `GenerationRequest` (used by `/responses`).
* **`mlx_vlm/generate.py`** — extend `generate_step` keyword-only signature with the four new parameters, plumb them through to `make_logits_processors`. Built the processor kwargs as an explicit dict so unset fields are *omitted* rather than passed as `None` (preserving mlx_lm's positional defaults). Updated docstrings on both `generate_step` and `generate`.
* **`mlx_vlm/tests/test_server.py`** — extend `test_chat_completions_endpoint_forwards_explicit_sampling_args` to assert the new fields reach `generate()`. (Could also extend `test_responses_endpoint_forwards_new_sampling_args` for symmetry — happy to add if you'd prefer.)

## Verification

Tested on Apple Silicon (Mac Mini, mlx-vlm installed from the patched branch) against `mlx-community/GLM-OCR-bf16`. With debug logging, all five new params are received correctly by `make_logits_processors`. Different combinations of the new params produce visibly different outputs at `temperature=0`, confirming they reach the sampler.

`repetition_penalty=2.0` alone gives a clean (1,188 chars) transcription on the previously-looping page. `frequency_penalty` combined with a milder `repetition_penalty` measurably reduces but does not fully eliminate looping at the values tested — so the new knob is a genuine tool, not a free lunch, but it's the one users currently can't reach.

## Compatibility

Pure addition. Existing requests that don't set the new fields behave identically. The only behavioural change is the new field appearing in the OpenAPI schema. No version bump required.

## Notes

Happy to iterate on:
- Whether to bump `repetition_penalty`'s default for known VLM model types (probably not — the failure surface differs by model).
- Whether to also add a small loop-detector / N-gram check (separate change, mentioned for completeness).
- Whether the `repetition_penalty` description copy is too opinionated — feel free to trim.

Thanks for maintaining this project. The Apple Silicon VLM ecosystem benefits enormously from it.

### CODE REVIEW

**Comment by lucasnewman** [comment]
Resolved in https://github.com/Blaizzy/mlx-vlm/pull/1214
